File: ios/OakApp/Features/Chat/SlashAutocomplete.swift
import SwiftUI

/// Composer `/` picker (SD-US-1). Shown above the composer in the same
/// neighborhood as ``MentionAutocomplete``. Tap inserts; Send still hops.
struct SlashAutocomplete: View {
  var commands: [SlashCommandRow] = []
  var names: [DexNameRow] = []
  var teams: [TeamSummary] = []
  var empty: String? = nil
  var isGuest: Bool = false
  var onPickCommand: (String) -> Void = { _ in }
  var onPickName: (DexNameRow) -> Void = { _ in }
  var onPickTeam: (TeamSummary) -> Void = { _ in }

  var body: some View {
    VStack(alignment: .leading, spacing: 0) {
      Text(SlashPicker.pickerCaption)
        .font(Theme.body(.caption))
        .foregroundStyle(Theme.textMuted)
        .padding(.horizontal, Theme.Spacing.md)
        .padding(.top, Theme.Spacing.sm)
        .padding(.bottom, Theme.Spacing.xs)

      ScrollView {
        LazyVStack(alignment: .leading, spacing: 0) {
          if !commands.isEmpty {
            ForEach(commands, id: \.token) { row in
              commandRow(row)
            }
          } else if !names.isEmpty {
            ForEach(names) { row in
              nameRow(row)
            }
          } else if !teams.isEmpty {
            ForEach(teams) { team in
              teamRow(team)
            }
          } else if let empty {
            Text(empty)
              .font(Theme.body(.subheadline))
              .foregroundStyle(Theme.textMuted)
              .padding(.horizontal, Theme.Spacing.md)
              .padding(.vertical, Theme.Spacing.sm)
          }
        }
      }
      .frame(maxHeight: 240)
    }
    .background(Theme.surface)
    .clipShape(RoundedRectangle(cornerRadius: Theme.Radius.md, style: .continuous))
    .overlay {
      RoundedRectangle(cornerRadius: Theme.Radius.md, style: .continuous)
        .strokeBorder(Theme.separator, lineWidth: 1)
    }
    .oakShadow(.raised)
    .accessibilityElement(children: .contain)
    .accessibilityLabel("Slash commands")
  }

  @ViewBuilder
  private func commandRow(_ row: SlashCommandRow) -> some View {
    let hint = isGuest ? (row.hintGuest ?? row.hint) : row.hint
    Button {
      onPickCommand(row.token)
    } label: {
      HStack {
        Text(row.token)
          .font(Theme.body(.subheadline, weight: .semibold))
          .foregroundStyle(Theme.textStrong)
        Spacer()
        Text(hint)
          .font(Theme.body(.caption))
          .foregroundStyle(Theme.textMuted)
          .multilineTextAlignment(.trailing)
      }
      .padding(.horizontal, Theme.Spacing.md)
      .padding(.vertical, Theme.Spacing.sm)
      .contentShape(Rectangle())
    }
    .buttonStyle(.plain)
    .accessibilityLabel("\(row.token), \(hint)")
  }

  @ViewBuilder
  private func nameRow(_ row: DexNameRow) -> some View {
    Button {
      onPickName(row)
    } label: {
      HStack {
        Text(row.displayName)
          .font(Theme.body(.subheadline))
          .foregroundStyle(Theme.textStrong)
        Spacer()
        Text(row.kind.label)
          .font(Theme.body(.caption))
          .foregroundStyle(Theme.textMuted)
      }
      .padding(.horizontal, Theme.Spacing.md)
      .padding(.vertical, Theme.Spacing.sm)
      .contentShape(Rectangle())
    }
    .buttonStyle(.plain)
    .accessibilityLabel("\(row.displayName), \(row.kind.label)")
  }

  @ViewBuilder
  private func teamRow(_ team: TeamSummary) -> some View {
    Button {
      onPickTeam(team)
    } label: {
      HStack {
        Text(team.name)
          .font(Theme.body(.subheadline))
          .foregroundStyle(Theme.textStrong)
        Spacer()
        Text(team.format.shortLabel)
          .font(Theme.body(.caption))
          .foregroundStyle(Theme.textMuted)
      }
      .padding(.horizontal, Theme.Spacing.md)
      .padding(.vertical, Theme.Spacing.sm)
      .contentShape(Rectangle())
    }
    .buttonStyle(.plain)
    .accessibilityLabel(team.name)
  }
}

File: android/app/src/main/kotlin/ai/gowtam/oak/features/chat/SlashAutocomplete.kt
package ai.gowtam.oak.features.chat

import ai.gowtam.oak.ui.LocalOakColors
import ai.gowtam.oak.ui.OakRadius
import ai.gowtam.oak.ui.OakSpacing
import ai.gowtam.oak.wire.TeamSummary
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp

/**
 * Composer `/` picker (SD-AC-1.1, SD-AC-8.6). Command rows, Dex/Usage name
 * rows, or an empty line, plus the insert caption. Tap inserts; the parent
 * hops on Send.
 */
@Composable
fun SlashAutocomplete(
    onPickCommand: (String) -> Unit,
    modifier: Modifier = Modifier,
    commands: List<SlashCommandRow> = emptyList(),
    names: List<DexNameRow> = emptyList(),
    teams: List<TeamSummary> = emptyList(),
    empty: String? = null,
    guest: Boolean = false,
    caption: String = PICKER_CAPTION,
    skipMove: Boolean = false,
    onPickName: (DexNameRow) -> Unit = {},
    onPickTeam: (TeamSummary) -> Unit = {},
    onSkipMove: () -> Unit = {},
) {
    val oak = LocalOakColors.current
    val shape = RoundedCornerShape(OakRadius.md)
    Column(
        modifier = modifier
            .fillMaxWidth()
            .testTag("slash-autocomplete")
            .semantics(mergeDescendants = false) {
                contentDescription = "Slash commands"
                role = Role.DropdownList
            }
            .background(MaterialTheme.colorScheme.surface, shape)
            .border(1.dp, oak.border, shape)
            .padding(vertical = OakSpacing.xs),
    ) {
        Text(
            text = caption,
            style = MaterialTheme.typography.bodySmall,
            color = oak.textMuted,
            modifier = Modifier.padding(horizontal = OakSpacing.md, vertical = OakSpacing.xs),
        )
        when {
            commands.isNotEmpty() -> {
                for (row in commands) {
                    val hint = if (guest) row.hintGuest ?: row.hint else row.hint
                    Text(
                        text = "${row.token}  $hint",
                        style = MaterialTheme.typography.bodyMedium,
                        color = oak.textStrong,
                        modifier = Modifier
                            .fillMaxWidth()
                            .semantics {
                                role = Role.Button
                                contentDescription = "${row.token} $hint"
                            }
                            .clickable { onPickCommand(row.token) }
                            .padding(horizontal = OakSpacing.md, vertical = OakSpacing.sm),
                    )
                }
            }
            skipMove || names.isNotEmpty() -> {
                if (skipMove) {
                    Row(
                        modifier = Modifier
                            .fillMaxWidth()
                            .testTag("slash-ac-skip-move")
                            .semantics {
                                role = Role.Button
                                contentDescription = "$CALC_SKIP_MOVE $CALC_SKIP_MOVE_HINT"
                            }
                            .clickable { onSkipMove() }
                            .padding(horizontal = OakSpacing.md, vertical = OakSpacing.sm),
                        verticalAlignment = Alignment.CenterVertically,
                    ) {
                        Text(
                            text = CALC_SKIP_MOVE,
                            style = MaterialTheme.typography.bodyMedium,
                            color = oak.textStrong,
                            modifier = Modifier.weight(1f),
                        )
                        Text(
                            text = CALC_SKIP_MOVE_HINT,
                            style = MaterialTheme.typography.bodySmall,
                            color = oak.textMuted,
                        )
                    }
                }
                for (row in names) {
                    Row(
                        modifier = Modifier
                            .fillMaxWidth()
                            .semantics {
                                role = Role.Button
                                contentDescription = "${row.displayName} ${row.kind.label()}"
                            }
                            .clickable { onPickName(row) }
                            .padding(horizontal = OakSpacing.md, vertical = OakSpacing.sm),
                        verticalAlignment = Alignment.CenterVertically,
                    ) {
                        Text(
                            text = row.displayName,
                            style = MaterialTheme.typography.bodyMedium,
                            color = oak.textStrong,
                            modifier = Modifier.weight(1f),
                        )
                        Text(
                            text = row.kind.label(),
                            style = MaterialTheme.typography.bodySmall,
                            color = oak.textMuted,
                        )
                    }
                }
            }
            teams.isNotEmpty() -> {
                for (team in teams) {
                    Text(
                        text = team.name,
                        style = MaterialTheme.typography.bodyMedium,
                        color = oak.textStrong,
                        modifier = Modifier
                            .fillMaxWidth()
                            .semantics { role = Role.Button }
                            .clickable { onPickTeam(team) }
                            .padding(horizontal = OakSpacing.md, vertical = OakSpacing.sm),
                    )
                }
            }
            empty != null -> {
                Text(
                    text = empty,
                    style = MaterialTheme.typography.bodySmall,
                    color = oak.textMuted,
                    modifier = Modifier.padding(horizontal = OakSpacing.md, vertical = OakSpacing.sm),
                )
            }
        }
    }
}
